In [2]:
import pandas as pd
import numpy as np
from datetime import datetime
from tqdm.notebook import tqdm 

<h6>Raw Public Data Source

In [3]:
csv_path_update = "../data/raw/>150K/public_150k_plus_230630.csv"

In [4]:
ppp_df_23 = pd.read_csv(csv_path_update)
ppp_df_23.head()

LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0  9547507704   05/01/2020            464              PPP   
1  9777677704   05/01/2020            464              PPP   
2  5791407702   05/01/2020           1013              PPP   
3  6223567700   05/01/2020            920              PPP   
4  9662437702   05/01/2020            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC                    NaN               NaN   

  BorrowerState BorrowerZip LoanStatusDate  ...             BusinessType  \
0           NaN  29150-9662     12/18/2020  ...              Corporation   
1           NaN  29420-9000     09/28/2021  ...      Sole Proprietorship   
2           NaN  98112-2922     03/17/2021  ...  Non-Profit Organization   
3           NaN  92407-1740     10/16/2021  ...              Corporation   
4           NaN         NaN     08/17/2021  ...                      NaN   

   OriginatingLenderLocationID                      OriginatingLender  \
0                        19248                           Synovus Bank   
1                        19248                           Synovus Bank   
2                         9551  Bank of America, National Association   
3                         9551  Bank of America, National Association   
4                        57328           The Huntington National Bank   

   OriginatingLenderCity  OriginatingLenderState      Gender      Veteran  \
0               COLUMBUS                      GA  Unanswered   Unanswered   
1               COLUMBUS                      GA  Male Owned  Non-Veteran   
2              CHARLOTTE                      NC  Unanswered   Unanswered   
3              CHARLOTTE                      NC  Male Owned  Non-Veteran   
4               COLUMBUS                      OH  Unanswered   Unanswered   

   NonProfit ForgivenessAmount ForgivenessDate  
0        NaN         773553.37      11/20/2020  
1        NaN         746336.24      08/12/2021  
2          Y         696677.49      02/10/2021  
3        NaN         395264.11      09/10/2021  
4        NaN         370819.35      04/08/2021  

[5 rows x 53 columns]

In [5]:
pd.set_option('display.max_rows', 60)
print(ppp_df_23.isnull().sum())
null_columns=ppp_df_23.columns[ppp_df_23.isnull().any()]
df_null = ppp_df_23[null_columns]
df_null

LoanNumber                          0
DateApproved                        0
SBAOfficeCode                       0
ProcessingMethod                    0
BorrowerName                        4
BorrowerAddress                    14
BorrowerCity                       12
BorrowerState                      13
BorrowerZip                        13
LoanStatusDate                  16825
LoanStatus                          0
Term                                0
SBAGuarantyPercentage               0
InitialApprovalAmount               0
CurrentApprovalAmount               0
UndisbursedAmount                  46
FranchiseName                  930147
ServicingLenderLocationID           0
ServicingLenderName                 0
ServicingLenderAddress              0
ServicingLenderCity                 0
ServicingLenderState                0
ServicingLenderZip                  0
RuralUrbanIndicator                 0
HubzoneIndicator                    0
LMIIndicator                        0
BusinessAgeD

BorrowerName  \
0                                   SUMTER COATINGS, INC.   
1                                   PLEASANT PLACES, INC.   
2                                 BOYER CHILDREN'S CLINIC   
3                                KIRTLEY CONSTRUCTION INC   
4                                            AERO BOX LLC   
...                                                   ...   
965547                           ROY E PAULSON, JR., P.C.   
965548  SWEETWATER COUNTY CHILD DEVELOPMENTAL CENTER, ...   
965549                  ELECTRICAL SYSTEMS OF WYOMING INC   
965550                                     EDEN LIFE CARE   
965551                      S & S JOHNSON ENTERPRISES INC   

                      BorrowerAddress      BorrowerCity BorrowerState  \
0               2410 Highway 15 South            Sumter           NaN   
1                 7684 Southrail Road  North Charleston           NaN   
2                    1850 BOYER AVE E           SEATTLE           NaN   
3                1661 MARTIN RANCH RD    SAN BERNARDINO           NaN   
4                                 NaN               NaN           NaN   
...                               ...               ...           ...   
965547                 102 N. Kenwood            CASPER            WY   
965548          1715 HITCHING POST DR       GREEN RIVER            WY   
965549                   1105 Adon Rd             Rozet            WY   
965550  30 N. Gould Street Suite 4000          SHERIDAN            WY   
965551           7342 Granite Loop Rd     TETON VILLAGE            WY   

       BorrowerZip LoanStatusDate  UndisbursedAmount FranchiseName  \
0       29150-9662     12/18/2020                0.0           NaN   
1       29420-9000     09/28/2021                0.0           NaN   
2       98112-2922     03/17/2021                0.0           NaN   
3       92407-1740     10/16/2021                0.0           NaN   
4              NaN     08/17/2021                0.0           NaN   
...            ...            ...                ...           ...   
965547  82601-2724     01/13/2021                0.0           NaN   
965548  82935-5783     12/08/2020                0.0           NaN   
965549  82727-8465     09/15/2021                0.0           NaN   
965550       82801     03/08/2022                0.0           NaN   
965551  83025-0550     10/22/2021                0.0           NaN   

                   BusinessAgeDescription       ProjectCity  ...  \
0       Existing or more than 2 years old            Sumter  ...   
1       Existing or more than 2 years old  North Charleston  ...   
2         New Business or 2 years or less           SEATTLE  ...   
3         New Business or 2 years or less    SAN BERNARDINO  ...   
4                              Unanswered               NaN  ...   
...                                   ...               ...  ...   
965547  Existing or more than 2 years old            CASPER  ...   
965548  Existing or more than 2 years old       GREEN RIVER  ...   
965549  Existing or more than 2 years old             Rozet  ...   
965550    New Business or 2 years or less          SHERIDAN  ...   
965551  Existing or more than 2 years old     TETON VILLAGE  ...   

       PAYROLL_PROCEED MORTGAGE_INTEREST_PROCEED RENT_PROCEED  \
0            769358.78                       NaN          NaN   
1            736927.79                       NaN          NaN   
2            691355.00                       NaN          NaN   
3            499871.00                       NaN          NaN   
4            367437.00                       NaN          NaN   
...                ...                       ...          ...   
965547       150000.00                       NaN          NaN   
965548       145000.00                       NaN          NaN   
965549       149997.00                       NaN          NaN   
965550       150000.00                       NaN          NaN   
965551       150000.00                       NaN          NaN 

<h5>Adding Loan Approval Amount Ranges to Determine Lender Percent Profit </h5>
https://home.treasury.gov/system/files/136/PPP%20Lender%20Information%20Fact%20Sheet.pdf

In [6]:
loan_conditions = [
(ppp_df_23.CurrentApprovalAmount <= 350000) & (ppp_df_23.CurrentApprovalAmount >= 150000),
(ppp_df_23.CurrentApprovalAmount <= 2000000) & (ppp_df_23.CurrentApprovalAmount > 350000),
(ppp_df_23.CurrentApprovalAmount > 2000000)
]
loan_values = [.05,.03,.01]
ppp_df_23['Estimated_LenderPercent'] = np.select(loan_conditions, loan_values)
ppp_df_23

LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0       9547507704   05/01/2020            464              PPP   
1       9777677704   05/01/2020            464              PPP   
2       5791407702   05/01/2020           1013              PPP   
3       6223567700   05/01/2020            920              PPP   
4       9662437702   05/01/2020            101              PPP   
...            ...          ...            ...              ...   
965547  4395967002   04/03/2020            897              PPP   
965548  6985647108   04/14/2020            897              PPP   
965549  7996438405   02/12/2021            897              PPS   
965550  9054647103   04/15/2020            897              PPP   
965551  9184687004   04/09/2020            897              PPP   

                                             BorrowerName  \
0                                   SUMTER COATINGS, INC.   
1                                   PLEASANT PLACES, INC.   
2                                 BOYER CHILDREN'S CLINIC   
3                                KIRTLEY CONSTRUCTION INC   
4                                            AERO BOX LLC   
...                                                   ...   
965547                           ROY E PAULSON, JR., P.C.   
965548  SWEETWATER COUNTY CHILD DEVELOPMENTAL CENTER, ...   
965549                  ELECTRICAL SYSTEMS OF WYOMING INC   
965550                                     EDEN LIFE CARE   
965551                      S & S JOHNSON ENTERPRISES INC   

                      BorrowerAddress      BorrowerCity BorrowerState  \
0               2410 Highway 15 South            Sumter           NaN   
1                 7684 Southrail Road  North Charleston           NaN   
2                    1850 BOYER AVE E           SEATTLE           NaN   
3                1661 MARTIN RANCH RD    SAN BERNARDINO           NaN   
4                                 NaN               NaN           NaN   
...                               ...               ...           ...   
965547                 102 N. Kenwood            CASPER            WY   
965548          1715 HITCHING POST DR       GREEN RIVER            WY   
965549                   1105 Adon Rd             Rozet            WY   
965550  30 N. Gould Street Suite 4000          SHERIDAN            WY   
965551           7342 Granite Loop Rd     TETON VILLAGE            WY   

       BorrowerZip LoanStatusDate  ... OriginatingLenderLocationID  \
0       29150-9662     12/18/2020  ...                       19248   
1       29420-9000     09/28/2021  ...                       19248   
2       98112-2922     03/17/2021  ...                        9551   
3       92407-1740     10/16/2021  ...                        9551   
4              NaN     08/17/2021  ...                       57328   
...            ...            ...  ...                         ...   
965547  82601-2724     01/13/2021  ...                       42366   
965548  82935-5783     12/08/2020  ...                      122813   
965549  82727-8465     09/15/2021  ...                       77189   
965550       82801     03/08/2022  ...                       25901   
965551  83025-0550     10/22/2021  ...                       77193   

                            OriginatingLender  OriginatingLenderCity  \
0                                Synovus Bank               COLUMBUS   
1                                Synovus Bank               COLUMBUS   
2       Bank of America, National Association              CHARLOTTE   
3       Bank of America, National Association              CHARLOTTE   
4                The Huntington National Bank               COLUMBUS   
...                                       ...                    ...   
965547                     Platte Valley Bank             TORRINGTON   
965548                      NebraskaLand Bank           NORTH PLATTE   
965549        First National Bank of Gillette               GILLETTE   
965550                    Small 

In [7]:
ppp_df_23["Estimated_LenderProfit"] = ppp_df_23['CurrentApprovalAmount'] * ppp_df_23['Estimated_LenderPercent']
ppp_df_23['Estimated_LenderPercent'] = pd.Series(["{0:.0f}%".format(val * 100) for val in ppp_df_23['Estimated_LenderPercent']], index = ppp_df_23.index)
ppp_df_23

LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0       9547507704   05/01/2020            464              PPP   
1       9777677704   05/01/2020            464              PPP   
2       5791407702   05/01/2020           1013              PPP   
3       6223567700   05/01/2020            920              PPP   
4       9662437702   05/01/2020            101              PPP   
...            ...          ...            ...              ...   
965547  4395967002   04/03/2020            897              PPP   
965548  6985647108   04/14/2020            897              PPP   
965549  7996438405   02/12/2021            897              PPS   
965550  9054647103   04/15/2020            897              PPP   
965551  9184687004   04/09/2020            897              PPP   

                                             BorrowerName  \
0                                   SUMTER COATINGS, INC.   
1                                   PLEASANT PLACES, INC.   
2                                 BOYER CHILDREN'S CLINIC   
3                                KIRTLEY CONSTRUCTION INC   
4                                            AERO BOX LLC   
...                                                   ...   
965547                           ROY E PAULSON, JR., P.C.   
965548  SWEETWATER COUNTY CHILD DEVELOPMENTAL CENTER, ...   
965549                  ELECTRICAL SYSTEMS OF WYOMING INC   
965550                                     EDEN LIFE CARE   
965551                      S & S JOHNSON ENTERPRISES INC   

                      BorrowerAddress      BorrowerCity BorrowerState  \
0               2410 Highway 15 South            Sumter           NaN   
1                 7684 Southrail Road  North Charleston           NaN   
2                    1850 BOYER AVE E           SEATTLE           NaN   
3                1661 MARTIN RANCH RD    SAN BERNARDINO           NaN   
4                                 NaN               NaN           NaN   
...                               ...               ...           ...   
965547                 102 N. Kenwood            CASPER            WY   
965548          1715 HITCHING POST DR       GREEN RIVER            WY   
965549                   1105 Adon Rd             Rozet            WY   
965550  30 N. Gould Street Suite 4000          SHERIDAN            WY   
965551           7342 Granite Loop Rd     TETON VILLAGE            WY   

       BorrowerZip LoanStatusDate  ...                      OriginatingLender  \
0       29150-9662     12/18/2020  ...                           Synovus Bank   
1       29420-9000     09/28/2021  ...                           Synovus Bank   
2       98112-2922     03/17/2021  ...  Bank of America, National Association   
3       92407-1740     10/16/2021  ...  Bank of America, National Association   
4              NaN     08/17/2021  ...           The Huntington National Bank   
...            ...            ...  ...                                    ...   
965547  82601-2724     01/13/2021  ...                     Platte Valley Bank   
965548  82935-5783     12/08/2020  ...                      NebraskaLand Bank   
965549  82727-8465     09/15/2021  ...        First National Bank of Gillette   
965550       82801     03/08/2022  ...                    Small Business Bank   
965551  83025-0550     10/22/2021  ...             Bank of Jackson Hole Trust   

        OriginatingLenderCity  OriginatingLenderState        Gender  \
0                    COLUMBUS                      GA    Unanswered   
1                    COLUMBUS                      GA    Male Owned   
2                   CHARLOTTE                      NC    Unanswered   
3                   CHARLOTTE                      NC    Male Owned   
4                    COLUMBUS                      OH    Unanswered   
...                       ...                     ...           ...   
965547             TORRINGTON                      WY    Male Owned   
965548           NORTH PLATTE                      N

<b>DATE TO TIMESTAMP CONVERSION 

DateApproved, ForgivenessDate

In [25]:
ppp_df_23.dtypes

LoanNumber            int64
SBAOfficeCode         int64
ProcessingMethod     object
BorrowerName         object
BorrowerAddress      object
                     ...   
4DigitNAICS         float64
NA                   object
Industry             object
SBAOfficeLabel       object
Loan_Range           object
Length: 63, dtype: object

In [12]:
dates = []
for date in ppp_df_23['DateApproved']:
    dt = pd.to_datetime(date, format='%m/%d/%Y') #year #### format
    dates.append(dt)

ppp_df_23['DateApprovedTS'] = dates
ppp_df_23['DateApprovedTS']= ppp_df_23['DateApprovedTS'].astype(str)
ppp_df_23['DateApprovedTS']= ppp_df_23['DateApprovedTS'].str.replace('-', '', regex=True).astype(int)
ppp_df_23.head()

LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0  9547507704   05/01/2020            464              PPP   
1  9777677704   05/01/2020            464              PPP   
2  5791407702   05/01/2020           1013              PPP   
3  6223567700   05/01/2020            920              PPP   
4  9662437702   05/01/2020            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC                    NaN               NaN   

  BorrowerState BorrowerZip LoanStatusDate  ... OriginatingLenderCity  \
0           NaN  29150-9662     12/18/2020  ...              COLUMBUS   
1           NaN  29420-9000     09/28/2021  ...              COLUMBUS   
2           NaN  98112-2922     03/17/2021  ...             CHARLOTTE   
3           NaN  92407-1740     10/16/2021  ...             CHARLOTTE   
4           NaN         NaN     08/17/2021  ...              COLUMBUS   

   OriginatingLenderState      Gender      Veteran  NonProfit  \
0                      GA  Unanswered   Unanswered        NaN   
1                      GA  Male Owned  Non-Veteran        NaN   
2                      NC  Unanswered   Unanswered          Y   
3                      NC  Male Owned  Non-Veteran        NaN   
4                      OH  Unanswered   Unanswered        NaN   

   ForgivenessAmount ForgivenessDate  Estimated_LenderPercent  \
0          773553.37      11/20/2020                       3%   
1          746336.24      08/12/2021                       3%   
2          696677.49      02/10/2021                       3%   
3          395264.11      09/10/2021                       3%   
4          370819.35      04/08/2021                       3%   

  Estimated_LenderProfit DateApprovedTS  
0             23080.7634       20200501  
1             22107.8337       20200501  
2             20740.6500       20200501  
3             14996.1300       20200501  
4             11023.1100       20200501  

[5 rows x 56 columns]

In [13]:
fdates = []
for fdate in ppp_df_23['ForgivenessDate']:
    fdt = pd.to_datetime(fdate, format='%m/%d/%Y') #year #### format
    fdates.append(fdt)

ppp_df_23['ForgivenessDateTS'] = fdates
ppp_df_23['ForgivenessDateTS']= ppp_df_23['ForgivenessDateTS'].astype(str)
ppp_df_23['ForgivenessDateTS']= ppp_df_23['ForgivenessDateTS'].str.replace('-', '', regex=True)
ppp_df_23.head()

LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0  9547507704   05/01/2020            464              PPP   
1  9777677704   05/01/2020            464              PPP   
2  5791407702   05/01/2020           1013              PPP   
3  6223567700   05/01/2020            920              PPP   
4  9662437702   05/01/2020            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC                    NaN               NaN   

  BorrowerState BorrowerZip LoanStatusDate  ... OriginatingLenderState  \
0           NaN  29150-9662     12/18/2020  ...                     GA   
1           NaN  29420-9000     09/28/2021  ...                     GA   
2           NaN  98112-2922     03/17/2021  ...                     NC   
3           NaN  92407-1740     10/16/2021  ...                     NC   
4           NaN         NaN     08/17/2021  ...                     OH   

       Gender      Veteran  NonProfit  ForgivenessAmount  ForgivenessDate  \
0  Unanswered   Unanswered        NaN          773553.37       11/20/2020   
1  Male Owned  Non-Veteran        NaN          746336.24       08/12/2021   
2  Unanswered   Unanswered          Y          696677.49       02/10/2021   
3  Male Owned  Non-Veteran        NaN          395264.11       09/10/2021   
4  Unanswered   Unanswered        NaN          370819.35       04/08/2021   

  Estimated_LenderPercent  Estimated_LenderProfit DateApprovedTS  \
0                      3%              23080.7634       20200501   
1                      3%              22107.8337       20200501   
2                      3%              20740.6500       20200501   
3                      3%              14996.1300       20200501   
4                      3%              11023.1100       20200501   

  ForgivenessDateTS  
0          20201120  
1          20210812  
2          20210210  
3          20210910  
4          20210408  

[5 rows x 57 columns]

In [14]:
ppp_df_23 = ppp_df_23.drop(['DateApproved', 'ForgivenessDate'], axis=1)
ppp_df_23.head()

LoanNumber  SBAOfficeCode ProcessingMethod              BorrowerName  \
0  9547507704            464              PPP     SUMTER COATINGS, INC.   
1  9777677704            464              PPP     PLEASANT PLACES, INC.   
2  5791407702           1013              PPP   BOYER CHILDREN'S CLINIC   
3  6223567700            920              PPP  KIRTLEY CONSTRUCTION INC   
4  9662437702            101              PPP              AERO BOX LLC   

         BorrowerAddress      BorrowerCity BorrowerState BorrowerZip  \
0  2410 Highway 15 South            Sumter           NaN  29150-9662   
1    7684 Southrail Road  North Charleston           NaN  29420-9000   
2       1850 BOYER AVE E           SEATTLE           NaN  98112-2922   
3   1661 MARTIN RANCH RD    SAN BERNARDINO           NaN  92407-1740   
4                    NaN               NaN           NaN         NaN   

  LoanStatusDate    LoanStatus  ...  OriginatingLenderCity  \
0     12/18/2020  Paid in Full  ...               COLUMBUS   
1     09/28/2021  Paid in Full  ...               COLUMBUS   
2     03/17/2021  Paid in Full  ...              CHARLOTTE   
3     10/16/2021  Paid in Full  ...              CHARLOTTE   
4     08/17/2021  Paid in Full  ...               COLUMBUS   

   OriginatingLenderState      Gender      Veteran  NonProfit  \
0                      GA  Unanswered   Unanswered        NaN   
1                      GA  Male Owned  Non-Veteran        NaN   
2                      NC  Unanswered   Unanswered          Y   
3                      NC  Male Owned  Non-Veteran        NaN   
4                      OH  Unanswered   Unanswered        NaN   

  ForgivenessAmount  Estimated_LenderPercent Estimated_LenderProfit  \
0         773553.37                       3%             23080.7634   
1         746336.24                       3%             22107.8337   
2         696677.49                       3%             20740.6500   
3         395264.11                       3%             14996.1300   
4         370819.35                       3%             11023.1100   

  DateApprovedTS ForgivenessDateTS  
0       20200501          20201120  
1       20200501          20210812  
2       20200501          20210210  
3       20200501          20210910  
4       20200501          20210408  

[5 rows x 55 columns]

In [15]:
pd.set_option('display.max_columns', None)
ppp_df_23.head()

LoanNumber  SBAOfficeCode ProcessingMethod              BorrowerName  \
0  9547507704            464              PPP     SUMTER COATINGS, INC.   
1  9777677704            464              PPP     PLEASANT PLACES, INC.   
2  5791407702           1013              PPP   BOYER CHILDREN'S CLINIC   
3  6223567700            920              PPP  KIRTLEY CONSTRUCTION INC   
4  9662437702            101              PPP              AERO BOX LLC   

         BorrowerAddress      BorrowerCity BorrowerState BorrowerZip  \
0  2410 Highway 15 South            Sumter           NaN  29150-9662   
1    7684 Southrail Road  North Charleston           NaN  29420-9000   
2       1850 BOYER AVE E           SEATTLE           NaN  98112-2922   
3   1661 MARTIN RANCH RD    SAN BERNARDINO           NaN  92407-1740   
4                    NaN               NaN           NaN         NaN   

  LoanStatusDate    LoanStatus  Term  SBAGuarantyPercentage  \
0     12/18/2020  Paid in Full    24                    100   
1     09/28/2021  Paid in Full    24                    100   
2     03/17/2021  Paid in Full    24                    100   
3     10/16/2021  Paid in Full    24                    100   
4     08/17/2021  Paid in Full    24                    100   

   InitialApprovalAmount  CurrentApprovalAmount  UndisbursedAmount  \
0              769358.78              769358.78                0.0   
1              736927.79              736927.79                0.0   
2              691355.00              691355.00                0.0   
3              499871.00              499871.00                0.0   
4              367437.00              367437.00                0.0   

  FranchiseName  ServicingLenderLocationID  \
0           NaN                      19248   
1           NaN                      19248   
2           NaN                       9551   
3           NaN                       9551   
4           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode  \
0           SC  29150-9662  SC-05          62.0   325510.0   
1           SC  29420-9000  SC-06          73.0   561730.0   
2           WA  98112-2922  WA-07          75.0        NaN   
3           CA  92407-1740  CA-23          21.0   236115.0   
4          NaN         NaN    NaN          25.0   484210.0   

                               Race               Ethnicity  \
0           

Determining NAICS for Industry of Each Loan

conditions = [
    (ppp_df_23['NA'] == 0),
    (ppp_df_23['NA'] == 99),
    (ppp_df_23['NA'] == 72),
    (ppp_df_23['NA'] == 56),
    (ppp_df_23['NA'] == 11),
    (ppp_df_23['NA'] == 71),
    (ppp_df_23['NA'] == 23),
    (ppp_df_23['NA'] == 61),
    (ppp_df_23['NA'] == 52),
    (ppp_df_23['NA'] == 62),
    (ppp_df_23['NA'] == 51),
    (ppp_df_23['NA'] == 55),
    (ppp_df_23['NA'] == 31),
    (ppp_df_23['NA'] == 32),
    (ppp_df_23['NA'] == 33),
    (ppp_df_23['NA'] == 21),
    (ppp_df_23['NA'] == 81),
    (ppp_df_23['NA'] == 54),
    (ppp_df_23['NA'] == 92),
    (ppp_df_23['NA'] == 53),
    (ppp_df_23['NA'] == 44),
    (ppp_df_23['NA'] == 45),
    (ppp_df_23['NA'] == 48),
    (ppp_df_23['NA'] == 49),
    (ppp_df_23['NA'] == 22),
    (ppp_df_23['NA'] == 42)
]

values = ['0',
 '99 - Nonclassifiable',
 'Accomodation and Food Services',
 'Administrative and Support and Waste Management and Remediation Services',
 'Agriculture, Forestry, Fishing and Hunting',
 'Arts, Entertainment, and Recreation',
 'Construction',
 'Educational Services',
 'Finance and Insurance',
 'Health Care and Social Assistance',
 'Information',
 'Management of Companies and Enterprises',
 'Manufacturing',
 'Manufacturing',
 'Manufacturing',
 'Mining',
 'Other Services (except Public Administration)',
 'Professional, Scientific, and Technical Services',
 'Public Administration',
 'Real Estate Rental and Leasing',
 'Retail Trade',
 'Retail Trade',
 'Transportation and Warehousing',
 'Transportation and Warehousing',
 'Utilities',
 'Wholesale Trade']

ppp_df_23['Industry'] = np.select(conditions, values)

ppp_df_23.head()

In [16]:
naics_ind = pd.read_csv("../data/alt_data/NAICS/NAICSind.csv") #Industry LongName
naics_full = pd.read_csv("../data/alt_data/NAICS/NAICSfull.csv") #SubIndustry

In [17]:
naics_ind["NA"] = naics_ind["NA"].astype(str)
naics_ind.head()

In [18]:
naics_full["NA"] = naics_full["NAICSCode"].astype(str).str[:2]
naics_full 

NAICSCode                                           LongName  \
0        111110                                    Soybean Farming   
1        111120                  Oilseed (except Soybean) Farming    
2        111130                          Dry Pea and Bean Farming    
3        111140                                      Wheat Farming   
4        111150                                      Corn Farming    
...         ...                                                ...   
1052     926140  Regulation of Agricultural Marketing and Commo...   
1053     926150  Regulation, Licensing, and Inspection of Misce...   
1054     927110                     Space Research and Technology    
1055     928110                                 National Security    
1056     928120                             International Affairs    

     TRIIndustrySector                                  IndustrySubsector  \
0            999 Other                     1111 Oilseed and Grain Farming   
1            999 Other                     1111 Oilseed and Grain Farming   
2            999 Other                     1111 Oilseed and Grain Farming   
3            999 Other                     1111 Oilseed and Grain Farming   
4            999 Other                     1111 Oilseed and Grain Farming   
...                ...                                                ...   
1052         999 Other          9261 Administration of Economic Programs    
1053         999 Other          9261 Administration of Economic Programs    
1054         999 Other                9271 Space Research and Technology    
1055         999 Other  9281 National Security and International Affairs    
1056         999 Other  9281 National Security and International Affairs    

      4DigitNAICS  NA  
0            1111  11  
1            1111  11  
2            1111  11  
3            1111  11  
4            1111  11  
...           ...  ..  
1052         9261  92  
1053         9261  92  
1054         9271  92  
1055         9281  92  
1056         9281  92  

[1057 rows x 6 columns]

In [20]:
naics_full = naics_full.merge(naics_ind, on ="NA", how="left")
naics_full

NAICSCode                                           LongName  \
0        111110                                    Soybean Farming   
1        111120                  Oilseed (except Soybean) Farming    
2        111130                          Dry Pea and Bean Farming    
3        111140                                      Wheat Farming   
4        111150                                      Corn Farming    
...         ...                                                ...   
1052     926140  Regulation of Agricultural Marketing and Commo...   
1053     926150  Regulation, Licensing, and Inspection of Misce...   
1054     927110                     Space Research and Technology    
1055     928110                                 National Security    
1056     928120                             International Affairs    

     TRIIndustrySector                                  IndustrySubsector  \
0            999 Other                     1111 Oilseed and Grain Farming   
1            999 Other                     1111 Oilseed and Grain Farming   
2            999 Other                     1111 Oilseed and Grain Farming   
3            999 Other                     1111 Oilseed and Grain Farming   
4            999 Other                     1111 Oilseed and Grain Farming   
...                ...                                                ...   
1052         999 Other          9261 Administration of Economic Programs    
1053         999 Other          9261 Administration of Economic Programs    
1054         999 Other                9271 Space Research and Technology    
1055         999 Other  9281 National Security and International Affairs    
1056         999 Other  9281 National Security and International Affairs    

      4DigitNAICS  NA                                    Industry  
0            1111  11  Agriculture, Forestry, Fishing and Hunting  
1            1111  11  Agriculture, Forestry, Fishing and Hunting  
2            1111  11  Agriculture, Forestry, Fishing and Hunting  
3            1111  11  Agriculture, Forestry, Fishing and Hunting  
4            1111  11  Agriculture, Forestry, Fishing and Hunting  
...           ...  ..                                         ...  
1052         9261  92                       Public Administration  
1053         9261  92                       Public Administration  
1054         9271  92                       Public Administration  
1055         9281  92                       Public Administration  
1056         9281  92                       Public Administration  

[1057 rows x 7 columns]

In [21]:
ppp_df_23 = ppp_df_23.merge(naics_full, on="NAICSCode", how="left")
ppp_df_23.head() #6632

LoanNumber  SBAOfficeCode ProcessingMethod              BorrowerName  \
0  9547507704            464              PPP     SUMTER COATINGS, INC.   
1  9777677704            464              PPP     PLEASANT PLACES, INC.   
2  5791407702           1013              PPP   BOYER CHILDREN'S CLINIC   
3  6223567700            920              PPP  KIRTLEY CONSTRUCTION INC   
4  9662437702            101              PPP              AERO BOX LLC   

         BorrowerAddress      BorrowerCity BorrowerState BorrowerZip  \
0  2410 Highway 15 South            Sumter           NaN  29150-9662   
1    7684 Southrail Road  North Charleston           NaN  29420-9000   
2       1850 BOYER AVE E           SEATTLE           NaN  98112-2922   
3   1661 MARTIN RANCH RD    SAN BERNARDINO           NaN  92407-1740   
4                    NaN               NaN           NaN         NaN   

  LoanStatusDate    LoanStatus  Term  SBAGuarantyPercentage  \
0     12/18/2020  Paid in Full    24                    100   
1     09/28/2021  Paid in Full    24                    100   
2     03/17/2021  Paid in Full    24                    100   
3     10/16/2021  Paid in Full    24                    100   
4     08/17/2021  Paid in Full    24                    100   

   InitialApprovalAmount  CurrentApprovalAmount  UndisbursedAmount  \
0              769358.78              769358.78                0.0   
1              736927.79              736927.79                0.0   
2              691355.00              691355.00                0.0   
3              499871.00              499871.00                0.0   
4              367437.00              367437.00                0.0   

  FranchiseName  ServicingLenderLocationID  \
0           NaN                      19248   
1           NaN                      19248   
2           NaN                       9551   
3           NaN                       9551   
4           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode  \
0           SC  29150-9662  SC-05          62.0   325510.0   
1           SC  29420-9000  SC-06          73.0   561730.0   
2           WA  98112-2922  WA-07          75.0        NaN   
3           CA  92407-1740  CA-23          21.0   236115.0   
4          NaN         NaN    NaN          25.0   484210.0   

                               Race               Ethnicity  \
0           

Determining SBA District Office for Processing Each Loan

SBA OFFICE CODE LABELS
0100	  	BOSTON REGIONAL OFFICE
0101	  	MASSACHUSETTS DISTRICT OFFICE
0130	  	SPRINGFIELD BRANCH OFFICE
0150	  	VERMONT DISTRICT OFFICE
0156	  	CONNECTICUT DISTRICT OFFICE
0165	  	RHODE ISLAND DISTRICT OFFICE
0172	  	MAINE DISTRICT OFFICE
0189	  	NEW HAMPSHIRE DISTRICT OFFICE
0200	  	NEW YORK REGIONAL OFFICE
0202	  	NEW YORK DISTRICT OFFICE
0206	  	ELMIRA BRANCH OFFICE
0219	  	ROCHESTER BRANCH OFFICE
0235	  	MELVILLE BRANCH OFFICE
0248	  	SYRACUSE DISTRICT OFFICE
0252	  	PUERTO RICO AND USVI DISTRICT OFFICE
0296	  	BUFFALO DISTRICT OFFICE
0299	  	NEW JERSEY DISTRICT OFFICE
0300	  	PHILADELPHIA REGIONAL OFFICE
0303	  	PHILADELPHIA DISTRICT OFFICE
0304	  	RICHMOND DISTRICT OFFICE
0316	  	HARRISBURG BRANCH OFFICE
0318	  	WILKES BARRE BRANCH OFFICE
0325	  	CHARLESTON BRANCH OFFICE
0341	  	DELAWARE DISTRICT OFFICE
0353	  	WASHINGTON DISTRICT OFFICE
0358	  	PITTSBURGH DISTRICT OFFICE
0373	  	BALTIMORE DISTRICT OFFICE
0390	  	WEST VIRGINIA DISTRICT OFFICE
0400	  	ATLANTA REGIONAL OFFICE
0405	  	GEORGIA DISTRICT OFFICE
0438	  	GULFPORT
0455	  	SOUTH FLORIDA DISTRICT OFFICE
0457	  	KENTUCKY DISTRICT OFFICE
0459	  	ALABAMA DISTRICT OFFICE
0460	  	NORTH CAROLINA DISTRICT OFFICE
0464	  	SOUTH CAROLINA DISTRICT OFFICE
0470	  	MISSISSIPPI DISTRICT OFFICE
0474	  	TENNESSEE DISTRICT OFFICE
0491	  	NORTH FLORIDA DISTRICT OFFICE
0500	  	CHICAGO REGIONAL OFFICE
0507	  	ILLINOIS DISTRICT OFFICE
0508	  	MINNESOTA DISTRICT OFFICE
0515	  	MICHIGAN DISTRICT OFFICE
0517	  	SPRINGFIELD BRANCH OFFICE
0545	  	CINCINNATI BRANCH OFFICE
0547	  	MARQUETTE BRANCH OFFICE
0549	  	CLEVELAND DISTRICT OFFICE
0562	  	INDIANA DISTRICT OFFICE
0563	  	WISCONSIN DISTRICT OFFICE
0593	  	COLUMBUS DISTRICT OFFICE
0600	  	DALLAS REGIONAL OFFICE
0610	  	DALLAS / FT WORTH DISTRICT OFFICE
0637	  	CORPUS CHRISTI BRANCH OFFICE
0639	  	LOWER RIO GRANDE VALLEY DISTRICT OFFICE
0669	  	ARKANSAS DISTRICT OFFICE
0671	  	HOUSTON DISTRICT OFFICE
0677	  	EL PASO DISTRICT OFFICE
0678	  	LUBBOCK DISTRICT OFFICE
0679	  	LOUISIANA DISTRICT OFFICE
0680	  	OKLAHOMA DISTRICT OFFICE
0681	  	SAN ANTONIO DISTRICT OFFICE
0682	  	NEW MEXICO DISTRICT OFFICE
0700	  	KANSAS CITY REGIONAL OFFICE
0709	  	KANSAS CITY DISTRICT OFFICE
0721	  	SPRINGFIELD BRANCH OFFICE
0736	  	CEDAR RAPIDS BRANCH OFFICE
0761	  	DES MOINES DISTRICT OFFICE
0766	  	NEBRASKA DISTRICT OFFICE
0767	  	WICHITA DISTRICT OFFICE
0768	  	ST. LOUIS DISTRICT OFFICE
0800	  	DENVER REGIONAL OFFICE
0811	  	COLORADO DISTRICT OFFICE
0875	  	NORTH DAKOTA DISTRICT OFFICE
0876	  	SOUTH DAKOTA DISTRICT OFFICE
0883	  	UTAH DISTRICT OFFICE
0885	  	MONTANA DISTRICT OFFICE
0897	  	WYOMING DISTRICT OFFICE
0900	  	SAN FRANCISCO REGIONAL OFFICE
0912	  	SAN FRANCISCO DISTRICT OFFICE
0914	  	LOS ANGELES DISTRICT OFFICE
0920	  	SANTA ANA DISTRICT OFFICE
0931	  	SACRAMENTO DISTRICT OFFICE
0942	  	FRESNO DISTRICT OFFICE
0944	  	NEVADA DISTRICT OFFICE
0951	  	HAWAII DISTRICT OFFICE
0954	  	SAN DIEGO DISTRICT OFFICE
0988	  	ARIZONA DISTRICT OFFICE
0995	  	GUAM BRANCH OFFICE
1000	  	SEATTLE REGIONAL OFFICE
1013	  	SEATTLE DISTRICT OFFICE
1084	  	ALASKA DISTRICT OFFICE
1086	  	PORTLAND DISTRICT OFFICE
1087	  	BOISE DISTRICT OFFICE
1094	  	SPOKANE BRANCH OFFICE

In [22]:
conditions = [
(ppp_df_23['SBAOfficeCode'] ==100),
(ppp_df_23['SBAOfficeCode'] ==101),
(ppp_df_23['SBAOfficeCode'] ==130),
(ppp_df_23['SBAOfficeCode'] ==150),
(ppp_df_23['SBAOfficeCode'] ==156),
(ppp_df_23['SBAOfficeCode'] ==165),
(ppp_df_23['SBAOfficeCode'] ==172),
(ppp_df_23['SBAOfficeCode'] ==189),
(ppp_df_23['SBAOfficeCode'] ==200),
(ppp_df_23['SBAOfficeCode'] ==202),
(ppp_df_23['SBAOfficeCode'] ==206),
(ppp_df_23['SBAOfficeCode'] ==219),
(ppp_df_23['SBAOfficeCode'] ==235),
(ppp_df_23['SBAOfficeCode'] ==248),
(ppp_df_23['SBAOfficeCode'] ==252),
(ppp_df_23['SBAOfficeCode'] ==296),
(ppp_df_23['SBAOfficeCode'] ==299),
(ppp_df_23['SBAOfficeCode'] ==300),
(ppp_df_23['SBAOfficeCode'] ==303),
(ppp_df_23['SBAOfficeCode'] ==304),
(ppp_df_23['SBAOfficeCode'] ==316),
(ppp_df_23['SBAOfficeCode'] ==318),
(ppp_df_23['SBAOfficeCode'] ==325),
(ppp_df_23['SBAOfficeCode'] ==341),
(ppp_df_23['SBAOfficeCode'] ==353),
(ppp_df_23['SBAOfficeCode'] ==358),
(ppp_df_23['SBAOfficeCode'] ==373),
(ppp_df_23['SBAOfficeCode'] ==390),
(ppp_df_23['SBAOfficeCode'] ==400),
(ppp_df_23['SBAOfficeCode'] ==405),
(ppp_df_23['SBAOfficeCode'] ==438),
(ppp_df_23['SBAOfficeCode'] ==455),
(ppp_df_23['SBAOfficeCode'] ==457),
(ppp_df_23['SBAOfficeCode'] ==459),
(ppp_df_23['SBAOfficeCode'] ==460),
(ppp_df_23['SBAOfficeCode'] ==464),
(ppp_df_23['SBAOfficeCode'] ==470),
(ppp_df_23['SBAOfficeCode'] ==474),
(ppp_df_23['SBAOfficeCode'] ==491),
(ppp_df_23['SBAOfficeCode'] ==500),
(ppp_df_23['SBAOfficeCode'] ==507),
(ppp_df_23['SBAOfficeCode'] ==508),
(ppp_df_23['SBAOfficeCode'] ==515),
(ppp_df_23['SBAOfficeCode'] ==517),
(ppp_df_23['SBAOfficeCode'] ==545),
(ppp_df_23['SBAOfficeCode'] ==547),
(ppp_df_23['SBAOfficeCode'] ==549),
(ppp_df_23['SBAOfficeCode'] ==562),
(ppp_df_23['SBAOfficeCode'] ==563),
(ppp_df_23['SBAOfficeCode'] ==593),
(ppp_df_23['SBAOfficeCode'] ==600),
(ppp_df_23['SBAOfficeCode'] ==610),
(ppp_df_23['SBAOfficeCode'] ==637),
(ppp_df_23['SBAOfficeCode'] ==639),
(ppp_df_23['SBAOfficeCode'] ==669),
(ppp_df_23['SBAOfficeCode'] ==671),
(ppp_df_23['SBAOfficeCode'] ==677),
(ppp_df_23['SBAOfficeCode'] ==678),
(ppp_df_23['SBAOfficeCode'] ==679),
(ppp_df_23['SBAOfficeCode'] ==680),
(ppp_df_23['SBAOfficeCode'] ==681),
(ppp_df_23['SBAOfficeCode'] ==682),
(ppp_df_23['SBAOfficeCode'] ==700),
(ppp_df_23['SBAOfficeCode'] ==709),
(ppp_df_23['SBAOfficeCode'] ==721),
(ppp_df_23['SBAOfficeCode'] ==736),
(ppp_df_23['SBAOfficeCode'] ==761),
(ppp_df_23['SBAOfficeCode'] ==766),
(ppp_df_23['SBAOfficeCode'] ==767),
(ppp_df_23['SBAOfficeCode'] ==768),
(ppp_df_23['SBAOfficeCode'] ==800),
(ppp_df_23['SBAOfficeCode'] ==811),
(ppp_df_23['SBAOfficeCode'] ==875),
(ppp_df_23['SBAOfficeCode'] ==876),
(ppp_df_23['SBAOfficeCode'] ==883),
(ppp_df_23['SBAOfficeCode'] ==885),
(ppp_df_23['SBAOfficeCode'] ==897),
(ppp_df_23['SBAOfficeCode'] ==900),
(ppp_df_23['SBAOfficeCode'] ==912),
(ppp_df_23['SBAOfficeCode'] ==914),
(ppp_df_23['SBAOfficeCode'] ==920),
(ppp_df_23['SBAOfficeCode'] ==931),
(ppp_df_23['SBAOfficeCode'] ==942),
(ppp_df_23['SBAOfficeCode'] ==944),
(ppp_df_23['SBAOfficeCode'] ==951),
(ppp_df_23['SBAOfficeCode'] ==954),
(ppp_df_23['SBAOfficeCode'] ==988),
(ppp_df_23['SBAOfficeCode'] ==995),
(ppp_df_23['SBAOfficeCode'] ==1000),
(ppp_df_23['SBAOfficeCode'] ==1013),
(ppp_df_23['SBAOfficeCode'] ==1084),
(ppp_df_23['SBAOfficeCode'] ==1086),
(ppp_df_23['SBAOfficeCode'] ==1087),
(ppp_df_23['SBAOfficeCode'] ==1094),
]

values = [
"BOSTON REGIONAL OFFICE",
"MASSACHUSETTS DISTRICT OFFICE",
"SPRINGFIELD BRANCH OFFICE",
"VERMONT DISTRICT OFFICE",
"CONNECTICUT DISTRICT OFFICE",
"RHODE ISLAND DISTRICT OFFICE",
"MAINE DISTRICT OFFICE",
"NEW HAMPSHIRE DISTRICT OFFICE",
"NEW YORK REGIONAL OFFICE",
"NEW YORK DISTRICT OFFICE",
"ELMIRA BRANCH OFFICE",
"ROCHESTER BRANCH OFFICE",
"MELVILLE BRANCH OFFICE",
"SYRACUSE DISTRICT OFFICE",
"PUERTO RICO AND USVI DISTRICT OFFICE",
"BUFFALO DISTRICT OFFICE",
"NEW JERSEY DISTRICT OFFICE",
"PHILADELPHIA REGIONAL OFFICE",
"PHILADELPHIA DISTRICT OFFICE",
"RICHMOND DISTRICT OFFICE",
"HARRISBURG BRANCH OFFICE",
"WILKES BARRE BRANCH OFFICE",
"CHARLESTON BRANCH OFFICE",
"DELAWARE DISTRICT OFFICE",
"WASHINGTON DISTRICT OFFICE",
"PITTSBURGH DISTRICT OFFICE",
"BALTIMORE DISTRICT OFFICE",
"WEST VIRGINIA DISTRICT OFFICE",
"ATLANTA REGIONAL OFFICE",
"GEORGIA DISTRICT OFFICE",
"GULFPORT",
"SOUTH FLORIDA DISTRICT OFFICE",
"KENTUCKY DISTRICT OFFICE",
"ALABAMA DISTRICT OFFICE",
"NORTH CAROLINA DISTRICT OFFICE",
"SOUTH CAROLINA DISTRICT OFFICE",
"MISSISSIPPI DISTRICT OFFICE",
"TENNESSEE DISTRICT OFFICE",
"NORTH FLORIDA DISTRICT OFFICE",
"CHICAGO REGIONAL OFFICE",
"ILLINOIS DISTRICT OFFICE",
"MINNESOTA DISTRICT OFFICE",
"MICHIGAN DISTRICT OFFICE",
"SPRINGFIELD BRANCH OFFICE",
"CINCINNATI BRANCH OFFICE",
"MARQUETTE BRANCH OFFICE",
"CLEVELAND DISTRICT OFFICE",
"INDIANA DISTRICT OFFICE",
"WISCONSIN DISTRICT OFFICE",
"COLUMBUS DISTRICT OFFICE",
"DALLAS REGIONAL OFFICE",
"DALLAS / FT WORTH DISTRICT OFFICE",
"CORPUS CHRISTI BRANCH OFFICE",
"LOWER RIO GRANDE VALLEY DISTRICT OFFICE",
"ARKANSAS DISTRICT OFFICE",
"HOUSTON DISTRICT OFFICE",
"EL PASO DISTRICT OFFICE",
"LUBBOCK DISTRICT OFFICE",
"LOUISIANA DISTRICT OFFICE",
"OKLAHOMA DISTRICT OFFICE",
"SAN ANTONIO DISTRICT OFFICE",
"NEW MEXICO DISTRICT OFFICE",
"KANSAS CITY REGIONAL OFFICE",
"KANSAS CITY DISTRICT OFFICE",
"SPRINGFIELD BRANCH OFFICE",
"CEDAR RAPIDS BRANCH OFFICE",
"DES MOINES DISTRICT OFFICE",
"NEBRASKA DISTRICT OFFICE",
"WICHITA DISTRICT OFFICE",
"ST. LOUIS DISTRICT OFFICE",
"DENVER REGIONAL OFFICE",
"COLORADO DISTRICT OFFICE",
"NORTH DAKOTA DISTRICT OFFICE",
"SOUTH DAKOTA DISTRICT OFFICE",
"UTAH DISTRICT OFFICE",
"MONTANA DISTRICT OFFICE",
"WYOMING DISTRICT OFFICE",
"SAN FRANCISCO REGIONAL OFFICE",
"SAN FRANCISCO DISTRICT OFFICE",
"LOS ANGELES DISTRICT OFFICE",
"SANTA ANA DISTRICT OFFICE",
"SACRAMENTO DISTRICT OFFICE",
"FRESNO DISTRICT OFFICE",
"NEVADA DISTRICT OFFICE",
"HAWAII DISTRICT OFFICE",
"SAN DIEGO DISTRICT OFFICE",
"ARIZONA DISTRICT OFFICE",
"GUAM BRANCH OFFICE",
"SEATTLE REGIONAL OFFICE",
"SEATTLE DISTRICT OFFICE",
"ALASKA DISTRICT OFFICE",
"PORTLAND DISTRICT OFFICE",
"BOISE DISTRICT OFFICE",
"SPOKANE BRANCH OFFICE"
]
ppp_df_23['SBAOfficeLabel'] = np.select(conditions, values)

ppp_df_23.head()

LoanNumber  SBAOfficeCode ProcessingMethod              BorrowerName  \
0  9547507704            464              PPP     SUMTER COATINGS, INC.   
1  9777677704            464              PPP     PLEASANT PLACES, INC.   
2  5791407702           1013              PPP   BOYER CHILDREN'S CLINIC   
3  6223567700            920              PPP  KIRTLEY CONSTRUCTION INC   
4  9662437702            101              PPP              AERO BOX LLC   

         BorrowerAddress      BorrowerCity BorrowerState BorrowerZip  \
0  2410 Highway 15 South            Sumter           NaN  29150-9662   
1    7684 Southrail Road  North Charleston           NaN  29420-9000   
2       1850 BOYER AVE E           SEATTLE           NaN  98112-2922   
3   1661 MARTIN RANCH RD    SAN BERNARDINO           NaN  92407-1740   
4                    NaN               NaN           NaN         NaN   

  LoanStatusDate    LoanStatus  Term  SBAGuarantyPercentage  \
0     12/18/2020  Paid in Full    24                    100   
1     09/28/2021  Paid in Full    24                    100   
2     03/17/2021  Paid in Full    24                    100   
3     10/16/2021  Paid in Full    24                    100   
4     08/17/2021  Paid in Full    24                    100   

   InitialApprovalAmount  CurrentApprovalAmount  UndisbursedAmount  \
0              769358.78              769358.78                0.0   
1              736927.79              736927.79                0.0   
2              691355.00              691355.00                0.0   
3              499871.00              499871.00                0.0   
4              367437.00              367437.00                0.0   

  FranchiseName  ServicingLenderLocationID  \
0           NaN                      19248   
1           NaN                      19248   
2           NaN                       9551   
3           NaN                       9551   
4           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode  \
0           SC  29150-9662  SC-05          62.0   325510.0   
1           SC  29420-9000  SC-06          73.0   561730.0   
2           WA  98112-2922  WA-07          75.0        NaN   
3           CA  92407-1740  CA-23          21.0   236115.0   
4          NaN         NaN    NaN          25.0   484210.0   

                               Race               Ethnicity  \
0           

In [23]:
ppp_df_23['SBAOfficeLabel'].nunique()

74

Determining Approval Ranges for Each Loan

In [24]:
loan_conditions = [
(ppp_df_23.CurrentApprovalAmount <= 350000) & (ppp_df_23.CurrentApprovalAmount >= 150000),
(ppp_df_23.CurrentApprovalAmount <= 1000000) & (ppp_df_23.CurrentApprovalAmount > 350000),
(ppp_df_23.CurrentApprovalAmount <= 2000000) & (ppp_df_23.CurrentApprovalAmount > 1000000),
(ppp_df_23.CurrentApprovalAmount <= 5000000) & (ppp_df_23.CurrentApprovalAmount > 2000000),
(ppp_df_23.CurrentApprovalAmount <= 10000000) & (ppp_df_23.CurrentApprovalAmount > 5000000),  
]
loan_values = ["a $150,000-350,000","b $350,000-1 million", "c $1-2 million","d $2-5 million", "e $5-10 million"]
ppp_df_23['Loan_Range'] = np.select(loan_conditions, loan_values)
ppp_df_23.head()

LoanNumber  SBAOfficeCode ProcessingMethod              BorrowerName  \
0  9547507704            464              PPP     SUMTER COATINGS, INC.   
1  9777677704            464              PPP     PLEASANT PLACES, INC.   
2  5791407702           1013              PPP   BOYER CHILDREN'S CLINIC   
3  6223567700            920              PPP  KIRTLEY CONSTRUCTION INC   
4  9662437702            101              PPP              AERO BOX LLC   

         BorrowerAddress      BorrowerCity BorrowerState BorrowerZip  \
0  2410 Highway 15 South            Sumter           NaN  29150-9662   
1    7684 Southrail Road  North Charleston           NaN  29420-9000   
2       1850 BOYER AVE E           SEATTLE           NaN  98112-2922   
3   1661 MARTIN RANCH RD    SAN BERNARDINO           NaN  92407-1740   
4                    NaN               NaN           NaN         NaN   

  LoanStatusDate    LoanStatus  Term  SBAGuarantyPercentage  \
0     12/18/2020  Paid in Full    24                    100   
1     09/28/2021  Paid in Full    24                    100   
2     03/17/2021  Paid in Full    24                    100   
3     10/16/2021  Paid in Full    24                    100   
4     08/17/2021  Paid in Full    24                    100   

   InitialApprovalAmount  CurrentApprovalAmount  UndisbursedAmount  \
0              769358.78              769358.78                0.0   
1              736927.79              736927.79                0.0   
2              691355.00              691355.00                0.0   
3              499871.00              499871.00                0.0   
4              367437.00              367437.00                0.0   

  FranchiseName  ServicingLenderLocationID  \
0           NaN                      19248   
1           NaN                      19248   
2           NaN                       9551   
3           NaN                       9551   
4           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode  \
0           SC  29150-9662  SC-05          62.0   325510.0   
1           SC  29420-9000  SC-06          73.0   561730.0   
2           WA  98112-2922  WA-07          75.0        NaN   
3           CA  92407-1740  CA-23          21.0   236115.0   
4          NaN         NaN    NaN          25.0   484210.0   

                               Race               Ethnicity  \
0           

Determining Job Report Ranges for each Loan

In [26]:
ppp_df_23["JobsReported"].dtype

dtype('float64')

In [27]:
ppp_df_23.loc[ppp_df_23["JobsReported"] == "0", 'JobsReported'] = 0

In [28]:
#possible use in NAICS division comparison, divisions based on Census of Employment and Wages - BLS - https://data.bls.gov/cew/apps/data_views/data_views.htm#tab=Tables
job_conditions = [
(ppp_df_23.JobsReported <= 5),
(ppp_df_23.JobsReported <= 9) & (ppp_df_23.JobsReported > 5),
(ppp_df_23.JobsReported <= 19) & (ppp_df_23.JobsReported > 9),
(ppp_df_23.JobsReported <= 49) & (ppp_df_23.JobsReported > 19),
(ppp_df_23.JobsReported <= 99) & (ppp_df_23.JobsReported > 49), 
(ppp_df_23.JobsReported <= 249) & (ppp_df_23.JobsReported > 99),
(ppp_df_23.JobsReported <= 499) & (ppp_df_23.JobsReported > 249),
(ppp_df_23.JobsReported > 499)
]
job_values = ["Fewer than 5","5 to 9", "10 to 19","20 to 49", "50 to 99", "100 to 249", "250 to 499", "500 or more"]
ppp_df_23['job_Range'] = np.select(job_conditions, job_values)
ppp_df_23.head()

LoanNumber  SBAOfficeCode ProcessingMethod              BorrowerName  \
0  9547507704            464              PPP     SUMTER COATINGS, INC.   
1  9777677704            464              PPP     PLEASANT PLACES, INC.   
2  5791407702           1013              PPP   BOYER CHILDREN'S CLINIC   
3  6223567700            920              PPP  KIRTLEY CONSTRUCTION INC   
4  9662437702            101              PPP              AERO BOX LLC   

         BorrowerAddress      BorrowerCity BorrowerState BorrowerZip  \
0  2410 Highway 15 South            Sumter           NaN  29150-9662   
1    7684 Southrail Road  North Charleston           NaN  29420-9000   
2       1850 BOYER AVE E           SEATTLE           NaN  98112-2922   
3   1661 MARTIN RANCH RD    SAN BERNARDINO           NaN  92407-1740   
4                    NaN               NaN           NaN         NaN   

  LoanStatusDate    LoanStatus  Term  SBAGuarantyPercentage  \
0     12/18/2020  Paid in Full    24                    100   
1     09/28/2021  Paid in Full    24                    100   
2     03/17/2021  Paid in Full    24                    100   
3     10/16/2021  Paid in Full    24                    100   
4     08/17/2021  Paid in Full    24                    100   

   InitialApprovalAmount  CurrentApprovalAmount  UndisbursedAmount  \
0              769358.78              769358.78                0.0   
1              736927.79              736927.79                0.0   
2              691355.00              691355.00                0.0   
3              499871.00              499871.00                0.0   
4              367437.00              367437.00                0.0   

  FranchiseName  ServicingLenderLocationID  \
0           NaN                      19248   
1           NaN                      19248   
2           NaN                       9551   
3           NaN                       9551   
4           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode  \
0           SC  29150-9662  SC-05          62.0   325510.0   
1           SC  29420-9000  SC-06          73.0   561730.0   
2           WA  98112-2922  WA-07          75.0        NaN   
3           CA  92407-1740  CA-23          21.0   236115.0   
4          NaN         NaN    NaN          25.0   484210.0   

                               Race               Ethnicity  \
0           

In [29]:
ppp_df_23.loc[ppp_df_23["job_Range"] == "0", 'job_Range'] = "Fewer than 5"
ppp_df_23['job_Range'].value_counts()

20 to 49        416149
10 to 19        241148
50 to 99        161163
100 to 249       84575
5 to 9           31080
250 to 499       21987
Fewer than 5      5093
500 or more       4357
Name: job_Range, dtype: int64

In [31]:
ppp_df_23['job_Range'].isnull().sum()

0

Combining Columns for Full Address 

In [32]:
ppp_df_23['full_add'] = ppp_df_23.BorrowerAddress + ", " + ppp_df_23.BorrowerCity + ", " + ppp_df_23.BorrowerState
ppp_df_23.head()

LoanNumber  SBAOfficeCode ProcessingMethod              BorrowerName  \
0  9547507704            464              PPP     SUMTER COATINGS, INC.   
1  9777677704            464              PPP     PLEASANT PLACES, INC.   
2  5791407702           1013              PPP   BOYER CHILDREN'S CLINIC   
3  6223567700            920              PPP  KIRTLEY CONSTRUCTION INC   
4  9662437702            101              PPP              AERO BOX LLC   

         BorrowerAddress      BorrowerCity BorrowerState BorrowerZip  \
0  2410 Highway 15 South            Sumter           NaN  29150-9662   
1    7684 Southrail Road  North Charleston           NaN  29420-9000   
2       1850 BOYER AVE E           SEATTLE           NaN  98112-2922   
3   1661 MARTIN RANCH RD    SAN BERNARDINO           NaN  92407-1740   
4                    NaN               NaN           NaN         NaN   

  LoanStatusDate    LoanStatus  Term  SBAGuarantyPercentage  \
0     12/18/2020  Paid in Full    24                    100   
1     09/28/2021  Paid in Full    24                    100   
2     03/17/2021  Paid in Full    24                    100   
3     10/16/2021  Paid in Full    24                    100   
4     08/17/2021  Paid in Full    24                    100   

   InitialApprovalAmount  CurrentApprovalAmount  UndisbursedAmount  \
0              769358.78              769358.78                0.0   
1              736927.79              736927.79                0.0   
2              691355.00              691355.00                0.0   
3              499871.00              499871.00                0.0   
4              367437.00              367437.00                0.0   

  FranchiseName  ServicingLenderLocationID  \
0           NaN                      19248   
1           NaN                      19248   
2           NaN                       9551   
3           NaN                       9551   
4           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode  \
0           SC  29150-9662  SC-05          62.0   325510.0   
1           SC  29420-9000  SC-06          73.0   561730.0   
2           WA  98112-2922  WA-07          75.0        NaN   
3           CA  92407-1740  CA-23          21.0   236115.0   
4          NaN         NaN    NaN          25.0   484210.0   

                               Race               Ethnicity  \
0           

Determining Business Age & Type Naming Consistency

In [33]:
ppp_df_23["BusinessType"] = ppp_df_23["BusinessType"].replace({'Limited  Liability Company(LLC)': 'Limited Liability Company(LLC)'})


In [34]:
ppp_df_23['BusinessAgeDescription'] = ppp_df_23['BusinessAgeDescription'].fillna('Unanswered')
ppp_df_23["BusinessType"] = ppp_df_23["BusinessType"].fillna('Not Available')
ppp_df_23["NonProfit"] = ppp_df_23["NonProfit"].fillna('N')
ppp_df_23.head()

LoanNumber  SBAOfficeCode ProcessingMethod              BorrowerName  \
0  9547507704            464              PPP     SUMTER COATINGS, INC.   
1  9777677704            464              PPP     PLEASANT PLACES, INC.   
2  5791407702           1013              PPP   BOYER CHILDREN'S CLINIC   
3  6223567700            920              PPP  KIRTLEY CONSTRUCTION INC   
4  9662437702            101              PPP              AERO BOX LLC   

         BorrowerAddress      BorrowerCity BorrowerState BorrowerZip  \
0  2410 Highway 15 South            Sumter           NaN  29150-9662   
1    7684 Southrail Road  North Charleston           NaN  29420-9000   
2       1850 BOYER AVE E           SEATTLE           NaN  98112-2922   
3   1661 MARTIN RANCH RD    SAN BERNARDINO           NaN  92407-1740   
4                    NaN               NaN           NaN         NaN   

  LoanStatusDate    LoanStatus  Term  SBAGuarantyPercentage  \
0     12/18/2020  Paid in Full    24                    100   
1     09/28/2021  Paid in Full    24                    100   
2     03/17/2021  Paid in Full    24                    100   
3     10/16/2021  Paid in Full    24                    100   
4     08/17/2021  Paid in Full    24                    100   

   InitialApprovalAmount  CurrentApprovalAmount  UndisbursedAmount  \
0              769358.78              769358.78                0.0   
1              736927.79              736927.79                0.0   
2              691355.00              691355.00                0.0   
3              499871.00              499871.00                0.0   
4              367437.00              367437.00                0.0   

  FranchiseName  ServicingLenderLocationID  \
0           NaN                      19248   
1           NaN                      19248   
2           NaN                       9551   
3           NaN                       9551   
4           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode  \
0           SC  29150-9662  SC-05          62.0   325510.0   
1           SC  29420-9000  SC-06          73.0   561730.0   
2           WA  98112-2922  WA-07          75.0        NaN   
3           CA  92407-1740  CA-23          21.0   236115.0   
4          NaN         NaN    NaN          25.0   484210.0   

                               Race               Ethnicity  \
0           

In [35]:
listbusiness =ppp_df_23["BusinessType"].value_counts()
listbusiness

Corporation                            415684
Limited Liability Company(LLC)         261058
Subchapter S Corporation               174799
Non-Profit Organization                 55904
Partnership                             18158
Limited Liability Partnership           12724
Sole Proprietorship                     12255
Professional Association                 6270
Cooperative                              2435
501(c)3 – Non Profit                     1772
Non-Profit Childcare Center               900
501(c)6 – Non Profit Membership           762
Not Available                             720
Employee Stock Ownership Plan(ESOP)       641
Self-Employed Individuals                 494
Trust                                     397
Housing Co-op                             152
Joint Venture                             125
Independent Contractors                   112
Single Member LLC                          78
Tribal Concerns                            64
Tenant in Common                  

Parsing State Files

In [37]:
state_counts = ppp_df_23["BorrowerState"].value_counts()
print(state_counts)

CA    129085
TX     76061
NY     73819
FL     60002
IL     40763
PA     38500
OH     32639
NJ     32515
MI     29791
MA     27179
GA     25787
WA     23619
VA     23224
NC     22796
MN     19720
CO     19669
MD     18955
WI     17346
IN     16594
MO     16339
TN     15864
AZ     15757
LA     13669
OR     13403
CT     12882
AL     10912
SC     10739
OK      9995
UT      9258
KY      9178
IA      8295
NV      8254
KS      8201
AR      5857
NE      5804
MS      5504
NH      5069
HI      4979
NM      4482
ID      4415
DC      4383
ME      4185
RI      3786
WV      3365
ND      3222
MT      3142
DE      2990
PR      2952
SD      2695
AK      2662
VT      2265
WY      2137
GU       443
VI       290
MP        82
AS        20
Name: BorrowerState, dtype: int64


In [38]:
ppp_df_23.to_csv("../data/ppp_pipe23.csv")

In [39]:
ak_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "AK", :]
al_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "AL", :]
ar_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "AR", :]
as_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "AS", :] #American Samoa
az_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "AZ", :]
ca_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "CA", :]
co_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "CO", :]
ct_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "CT", :]
dc_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "DC", :] #D.C.
de_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "DE", :]
fl_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "FL", :]
ga_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "GA", :]
gu_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "GU", :] #Guam
hi_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "HI", :]
ia_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "IA", :]
id_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "ID", :]
il_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "IL", :]
in_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "IN", :]
ks_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "KS", :]
ky_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "KY", :]
la_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "LA", :]
ma_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "MA", :]
md_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "MD", :]
me_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "ME", :]
mi_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "MI", :]
mn_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "MN", :]
mo_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "MO", :]
mp_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "MP", :] #Saipan Member of Parliament
ms_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "MS", :]
mt_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "MT", :]
nc_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "NC", :]
nd_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "ND", :]
ne_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "NE", :]
nh_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "NH", :]
nj_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "NJ", :]
nm_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "NM", :]
nv_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "NV", :]
ny_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "NY", :]
oh_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "OH", :]
ok_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "OK", :]
or_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "OR", :]
pa_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "PA", :]
pr_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "PR", :] #Puerto Rico
ri_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "RI", :]
sc_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "SC", :]
sd_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "SD", :]
tn_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "TN", :]
tx_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "TX", :]
ut_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "UT", :]
va_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "VA", :]
vi_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "VI", :] #Virgin Islands
vt_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "VT", :]
wa_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "WA", :]
wi_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "WI", :]
wv_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "WV", :]
wy_df = ppp_df_23.loc[ppp_df_23["BorrowerState"] == "WY", :]

In [40]:
ak_df.to_csv("../data/state_data/pipe_states/23/ak.csv") 
al_df.to_csv("../data/state_data/pipe_states/23/al.csv")
ar_df.to_csv("../data/state_data/pipe_states/23/ar.csv") 
as_df.to_csv("../data/state_data/pipe_states/23/as.csv")
az_df.to_csv("../data/state_data/pipe_states/23/az.csv")  
ca_df.to_csv("../data/state_data/pipe_states/23/ca.csv")  
co_df.to_csv("../data/state_data/pipe_states/23/co.csv") 
ct_df.to_csv("../data/state_data/pipe_states/23/ct.csv")  
dc_df.to_csv("../data/state_data/pipe_states/23/dc.csv")
de_df.to_csv("../data/state_data/pipe_states/23/de.csv") 
fl_df.to_csv("../data/state_data/pipe_states/23/fl.csv") 
ga_df.to_csv("../data/state_data/pipe_states/23/ga.csv") 
gu_df.to_csv("../data/state_data/pipe_states/23/gu.csv")
hi_df.to_csv("../data/state_data/pipe_states/23/hi.csv") 
ia_df.to_csv("../data/state_data/pipe_states/23/ia.csv")  
id_df.to_csv("../data/state_data/pipe_states/23/id.csv")  
il_df.to_csv("../data/state_data/pipe_states/23/il.csv")  
in_df.to_csv("../data/state_data/pipe_states/23/in.csv") 
ks_df.to_csv("../data/state_data/pipe_states/23/ks.csv")  
ky_df.to_csv("../data/state_data/pipe_states/23/ky.csv")  
la_df.to_csv("../data/state_data/pipe_states/23/la.csv")
ma_df.to_csv("../data/state_data/pipe_states/23/ma.csv") 
md_df.to_csv("../data/state_data/pipe_states/23/md.csv")
me_df.to_csv("../data/state_data/pipe_states/23/me.csv")  
mi_df.to_csv("../data/state_data/pipe_states/23/mi.csv") 
mn_df.to_csv("../data/state_data/pipe_states/23/mn.csv") 
mo_df.to_csv("../data/state_data/pipe_states/23/mo.csv")  
mp_df.to_csv("../data/state_data/pipe_states/23/mp.csv") 
ms_df.to_csv("../data/state_data/pipe_states/23/ms.csv")
mt_df.to_csv("../data/state_data/pipe_states/23/mt.csv")
nc_df.to_csv("../data/state_data/pipe_states/23/nc.csv") 
nd_df.to_csv("../data/state_data/pipe_states/23/nd.csv")  
ne_df.to_csv("../data/state_data/pipe_states/23/ne.csv")
nh_df.to_csv("../data/state_data/pipe_states/23/nh.csv")  
nj_df.to_csv("../data/state_data/pipe_states/23/nj.csv")
nm_df.to_csv("../data/state_data/pipe_states/23/nm.csv") 
nv_df.to_csv("../data/state_data/pipe_states/23/nv.csv")  
ny_df.to_csv("../data/state_data/pipe_states/23/ny.csv")  
oh_df.to_csv("../data/state_data/pipe_states/23/oh.csv")
ok_df.to_csv("../data/state_data/pipe_states/23/ok.csv") 
or_df.to_csv("../data/state_data/pipe_states/23/or.csv")  
pa_df.to_csv("../data/state_data/pipe_states/23/pa.csv")
pr_df.to_csv("../data/state_data/pipe_states/23/pr.csv")  
ri_df.to_csv("../data/state_data/pipe_states/23/ri.csv") 
sc_df.to_csv("../data/state_data/pipe_states/23/sc.csv") 
sd_df.to_csv("../data/state_data/pipe_states/23/sd.csv") 
tn_df.to_csv("../data/state_data/pipe_states/23/tn.csv")  
tx_df.to_csv("../data/state_data/pipe_states/23/tx.csv")  
ut_df.to_csv("../data/state_data/pipe_states/23/ut.csv")  
va_df.to_csv("../data/state_data/pipe_states/23/va.csv")  
vi_df.to_csv("../data/state_data/pipe_states/23/vi.csv")  
vt_df.to_csv("../data/state_data/pipe_states/23/vt.csv")
wa_df.to_csv("../data/state_data/pipe_states/23/wa.csv")  
wi_df.to_csv("../data/state_data/pipe_states/23/wi.csv")  
wv_df.to_csv("../data/state_data/pipe_states/23/wv.csv") 
wy_df.to_csv("../data/state_data/pipe_states/23/wy.csv")  

Possible Additions to Pipeline:<br>
Location-based:    http://www.city-data.com/ <br>
Dates:
using deep learning with time series data is to split the date into multiple categorical variables (year, month, week, day of week, day of month, and Booleans for whether it’s the start/end of a month/quarter/year). 